File: Data Structures & Algorithms/first-missing-positive/submission-0.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        arr = [False]*(n)

        for i in range(n):
            if 1 <= nums[i] <= n:
                arr[nums[i] - 1] = True
        
        for i in range(n):
            if not arr[i]:
                return i + 1
        return n + 1
         

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> nums;
    int target;

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        this->nums = candidates;    
        this->target = target;
        sort(nums.begin(), nums.end());

        vector<int> state;
        foo(0, 0, state);

        return res;
    }

    void foo(int currentSum, int idx, vector<int>& state) {
        if(currentSum == target) {
            this->res.push_back(state);
            return;
        }

        for(int i = idx; i < nums.size(); i++) {
            int n = nums[i];
            if (i > idx && nums[i] == nums[i-1]) continue;
            if(currentSum <= target) {
                state.push_back(n);
                foo(currentSum + n, i + 1, state);
                state.pop_back();
            }
        }
    }
};
